clang/test/SemaSYCL/sycl-device-template-diag.cpp

```diff
@@ -0,0 +1,36 @@
+// RUN: %clang_cc1 -fsycl-is-device -verify -Wno-sycl-2017-compat -fsyntax-only %s -internal-isystem %S/Inputs
+
+// This test verifies that we generate deferred diagnostics when
+// such diagnostics are in a function template.
+
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+queue q;
+template <typename h>
+h *malloc_shared();
+
+int global_value = -1;
+
+template <typename T>
+void kernel_wrapper() {
+  q.submit([&](handler &h) {
+    h.single_task([=] {
+      // expected-error@+1{{SYCL kernel cannot use a non-const global variable}}
+      (void)global_value;
+    });
+  });
+}
+
+int main() {
+  int *x = malloc_shared<int>();
+  kernel_handler kh;
+  q.submit([&](handler &h) {
+    h.single_task<class mykern>([=](auto g) {
+      // expected-error@+1{{SYCL kernel cannot use a non-const global variable}}
+      x[3] = global_value;
+    },
+                                kh);
+  });
```

Comment:
NIT: If the intention of this test is to check how deferred diagnostics work with **templates**, I think we can eliminate this part.
```suggestion
```

Reply:
This is true.  I had added the non-template part of it to show that that case was already diagnosed, but now that we diagnose the template case also, this can be removed.   Will do soon.